tools/comp/markv.cpp

```diff
@@ -0,0 +1,229 @@
+
+USAGE: %s [e|d] [options] [<filename>]
+
+The input binary is read from <filename>. If no file is specified,
+or if the filename is "-", then the binary is read from standard input.
+
+If no output is specified then the output is printed to stdout in a human
+readable format.
+
+Tasks:
+  e               Encode SPIR-V to MARK-V.
+  d               Decode MARK-V to SPIR-V.
+
+Options:
+  -h, --help      Print this help.
+  --comments      Write codec comments to stdout.
+  --version       Display MARK-V codec version.
+
+  -o <filename>   Set the output filename.
+                  Output goes to standard output if this option is
+                  not specified, or if the filename is "-".
+)",
+      argv0, argv0);
+}
+
+void DiagnosticsMessageHandler(spv_message_level_t level, const char*,
+                               const spv_position_t& position,
+                               const char* message) {
+  switch (level) {
+    case SPV_MSG_FATAL:
+    case SPV_MSG_INTERNAL_ERROR:
+    case SPV_MSG_ERROR:
+      std::cerr << "error: " << position.index << ": " << message
+                << std::endl;
+      break;
+    case SPV_MSG_WARNING:
+      std::cout << "warning: " << position.index << ": " << message
+                << std::endl;
+      break;
+    case SPV_MSG_INFO:
+      std::cout << "info: " << position.index << ": " << message << std::endl;
+      break;
+    default:
+      break;
+  }
+}
+
+}  // namespace
+
+int main(int argc, char** argv) {
+  const char* input_filename = nullptr;
+  const char* output_filename = nullptr;
+
+  Task task = kNoTask;
+
+  bool want_comments = false;
+
+  for (int argi = 1; argi < argc; ++argi) {
+    if ('-' == argv[argi][0]) {
+      switch (argv[argi][1]) {
+        case 'h':
+          print_usage(argv[0]);
+          return 0;
+        case 'o': {
+          if (!output_filename && argi + 1 < argc) {
+            output_filename = argv[++argi];
+          } else {
+            print_usage(argv[0]);
+            return 1;
+          }
+        } break;
+        case '-': {
+          if (0 == strcmp(argv[argi], "--help")) {
+            print_usage(argv[0]);
+            return 0;
+          } else if (0 == strcmp(argv[argi], "--comments")) {
+            want_comments = true;
+          } else if (0 == strcmp(argv[argi], "--version")) {
+            fprintf(stderr, "error: Not implemented\n");
+            return 1;
+          } else {
+            print_usage(argv[0]);
+            return 1;
+          }
+        } break;
+        case '\0': {
+          // Setting a filename of "-" to indicate stdin.
+          if (!input_filename) {
+            input_filename = argv[argi];
+          } else {
+            fprintf(stderr, "error: More than one input file specified\n");
+            return 1;
+          }
+        } break;
+        default:
+          print_usage(argv[0]);
+          return 1;
+      }
+    } else if (0 == strcmp("e", argv[argi])) {
+      task = kEncode;
```

Comment:
What if the filename is "e"  ?

Then you'll never pick it up.

E.g. spirv-markv e e